### DISCUSSION TITLE
Quantum import convention

### DISCUSSION THREAD

**Discussion by ss2165** [Quantinuum/guppylang#641] [Ideas]
Currently guppy examples import all quantuim functions at the top of the python module. Something like the following excerpt from the [adaptive qpe example](https://github.com/CQCL-DEV/guppy-integration/blob/main/notebooks/adaptive-qpe.ipynb):

```python
from guppylang import guppy
from guppylang.prelude.angles import angle
from guppylang.prelude.builtins import py, result
from guppylang.prelude.quantum import discard, measure, qubit, h, rz, x, zz_phase, crz

@guppy
def oracle(ctrl: qubit, q: qubit, t: float) -> None:
    """Applies a controlled e^-iπHt/2 gate for the example Hamiltonian H = 0.5 * Z."""
    crz(ctrl, q, angle(0.5*t))

@guppy
def eigenstate() -> qubit:
    """Prepares eigenstate of the example Hamiltonian H = 0.5 * Z."""
    q = qubit()
    x(q)
    return q
```

Having to go back and import so many things from one place is not nice because:
1. Source of frustration every time you want a new op
2. No IDE auto complete for operations at point of use
3. Long import lines

It is for these kinds of reasons that many python libraries with a large range of top level functions and types use the import abbreviation convention:

```python
import numpy as np
import pandas as pd
import tensorflow as tf
import qnexus as qnx
```

I propose we adopt/promote this in guppy, but focussed on the standard library. You should also be able to conveniently import built-ins (ones that don't clash with python built-ins) without having to import them one by one. This relates to #430.  So rewriting the example above:

```python
from guppylang import guppy
from guppylang.prelude import * # brings things like `py`, `result`, `qubit`, `array` in to python scope
from guppylang.std import quantum as gq # better two letter abbreviations welcome

@guppy
def oracle(ctrl: qubit, q: qubit, t: float) -> None:
    """Applies a controlled e^-iπHt/2 gate for the example Hamiltonian H = 0.5 * Z."""
    gq.crz(ctrl, q, angle(0.5*t)) # I got `crz` suggested in my IDE when I typed `gq.`!

@guppy
def eigenstate() -> qubit:
    """Prepares eigenstate of the example Hamiltonian H = 0.5 * Z."""
    q = qubit()
    gq.x(q)
    return q
```

(Sneaky bonus: https://github.com/CQCL/guppylang/issues/640)

**Comment by acl-cqc** [comment]
This is all about convention in examples, not what the language supports (which just mirrors python), right?

**Reply by ss2165** [reply]
yes

**Comment by dunc-go** [comment]
I don't love that 
`from guppylang.std.builtins import control, dagger` etc
doesn't really show these are some grouped special compatible class of 'modifiers' -- they kind of look like they are gates (but @nicolaassolini-qntm pointed out to me that gates are from guppylang.std.quantum

my immediate preference is guppylang.std.modifiers (which is probably too small in scope) and I defer to wiser folks
